src/utils.ts

```diff
@@ -303,3 +313,48 @@ export async function createUnsignedFillRelayTransaction(
     { from: simulatedRelayerAddress }
   );
 }
+
+/**
+ * Search back in time for the first block at or older than the desired lookback.
+ *
+ * The approach here is as below:
+ * 1. We use the skip distances, which are meant to be a couple of hours or less for each chain.
+ * 2. Fetch the first block one skip distance from the latest block. Check its timestamp to see how far back it is
+ * relatively to the latest block.
+ * 3. Use the implied block speed from the latest block to one skip distance earlier to approximate how many blocks
+ * earlier we need to lookback to find a block older than the target block. This assumes that the block speed is roughly
+ * constant from latest block until the target block.
+ * 4. Check the timestamp again, if we still haven't achieved the desired look back, repeat steps 2-3 and go back
+ * further. Otherwise, return the block number.
+ *
+ * This is only a rough estimate and can go back further than desired. However, this would minimize the number
+ * of requests and is generally accurate unless the rate of block production has been very volatile over a
+ * short period of time.
+ *
+ * @param provider A valid provider - will be used to find the block with desired lookback.
+ * @param desiredLookback Desired lookback in seconds (e.g. 86400 seconds or 1 day).
+ * @return The block number that's at or older than the desired lookback.
+ */
+export async function findBlockAtOrOlder(provider: providers.Provider, desiredLookback: number): Promise<number> {
+  const [toBlock, network] = await Promise.all([provider.getBlock("latest"), provider.getNetwork()]);
+  let toBlockTimestamp = toBlock.timestamp;
+  const desiredTimestamp = toBlockTimestamp - desiredLookback;
```

Comment:
nit: suggest asserting that `desiredLookback <= toBlockTimestamp`

Reply:
+1

Reply:
Updated